### PR TITLE
Backport claim assets tests from polkadot-fellows repo

### PR DESCRIPTION
**Pull request by programskillforverification**
# Issue
[Backport integration tests for claim assets to the polkadot-sdk from polkadot-fellows repo #4892](https://github.com/paritytech/polkadot-sdk/issues/4892)
# Description
For the first time to contribute this project, I just finished part and make sure everything is ok.
- [x] backport [claim_assets test case](https://github.com/polkadot-fellows/runtimes/blob/main/integration-tests/emulated/tests/assets/asset-hub-kusama/src/tests/claim_assets.rs) to the polkadot-sdk testnet integration tests
- [x] backport macro [test_chain_can_claim_assets](https://github.com/polkadot-fellows/runtimes/blob/main/integration-tests/emulated/helpers/src/lib.rs#L218) from fellows repo
- [ ] when merged to polkadot-sdk and released, make sure that it is propagated to the fellows repo: https://github.com/polkadot-fellows/runtimes/issues/363
- [x] backport and align other macros/test-cases from https://github.com/polkadot-fellows/runtimes/blob/8ec28f96eeb30fbba30d29006d75e1a3fa1cea1c/integration-tests/emulated/helpers/src/lib.rs#L31-L33

### CODE REVIEW

File: cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/mod.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+mod claim_assets;
```
Comment (bkontur) [inline cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/mod.rs:16]: @programskillforverification can you please add this test for other runtime tests under `cumulus/parachains/integration-tests/emulated/tests/` directory?

Comment (programskillforverification) [reply cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/mod.rs:16]: @bkontur Yes. I just want to run the check and find issues. I will do other parts right now.

File: bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs [hunk 2]
```diff
@@ -664,7 +664,8 @@ impl<'a, P: SubstrateParachainsPipeline, SourceRelayClnt, TargetClnt>
 	for (
 		&'a OnDemandParachainsRelay<P, SourceRelayClnt, TargetClnt>,
 		&'a ParachainsSource<P, SourceRelayClnt>,
-	) where
+	)
+where
```
[the remark this answers is not in the document]
Comment (programskillforverification) [reply bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs:668]: @bkontur in this commit, I run `cargo +nightly fmt` so the formatting change. Now I run `cargo +nightly fmt --all -- --check` but nothing happen. I think maybe current formatting is correct.

[the remark this answers is not in the document]
Comment (bkontur) [reply bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs:668]: > @bkontur in this commit, I run `cargo +nightly fmt` so the formatting change. Now I run `cargo +nightly fmt --all -- --check` but nothing happen. I think maybe current formatting is correct.

well, it is not :)
https://github.com/paritytech/polkadot-sdk/actions/runs/9780959031/job/27006106808?pr=4930
what nightly version do you use?

[the remark this answers is not in the document]
Comment (programskillforverification) [reply bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs:668]: > well, it is not :) https://github.com/paritytech/polkadot-sdk/actions/runs/9780959031/job/27006106808?pr=4930 what nightly version do you use?

1.81.0-nightly

**Comment by bkontur** [comment]
bot fmt

File: bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkontur) [reply bridges/relays/lib-substrate-relay/src/on_demand/parachains.rs]: CI uses:
```
cargo 1.79.0-nightly (28e7b2bc0 2024-04-05)
rustc 1.79.0-nightly (8b2459c1f 2024-04-09)
```
I fixed formatting with our bot

**Review by franciscoaguirre — COMMENTED** [review]
We probably want the test for claiming assets on all runtimes, thinking people chain and coretime chain for sure

**Comment by programskillforverification** [comment]
> We probably want the test for claiming assets on all runtimes, thinking people chain and coretime chain for sure

@franciscoaguirre Okay, claiming assets will be added to people chain. But there is no any tests about coretime chain in whole integration tests directory. Are you sure to import it?

**Comment by acatangiu** [comment]
> But there is no any tests about coretime chain in whole integration tests directory. Are you sure to import it?

Yes. We should add emulated tests for coretime chain as well.

**Comment by programskillforverification** [comment]
I just add tests in rococo. If everything is ok, especially config, I will add them in westend.

File: Cargo.toml [hunk 3]
```diff
@@ -667,6 +671,7 @@ colored = { version = "2.0.4" }
 comfy-table = { version = "7.1.0", default-features = false }
 console = { version = "0.15.8" }
 contracts-rococo-runtime = { path = "cumulus/parachains/runtimes/contracts/contracts-rococo" }
+coretime-rococo-emulated-chain = { path = "cumulus/parachains/integration-tests/emulated/chains/parachains/coretime/coretime-rococo" }
 coretime-rococo-runtime = { path = "cumulus/parachains/runtimes/coretime/coretime-rococo" }
 coretime-westend-runtime = { path = "cumulus/parachains/runtimes/coretime/coretime-westend" }
```
Comment (acatangiu) [inline Cargo.toml:676]: Should there not be `coretime-westend-emulated-chain` here?

```suggestion
coretime-rococo-emulated-chain = { path = "cumulus/parachains/integration-tests/emulated/chains/parachains/coretime/coretime-rococo" }
coretime-rococo-runtime = { path = "cumulus/parachains/runtimes/coretime/coretime-rococo" }
coretime-westend-emulated-chain = { path = "cumulus/parachains/integration-tests/emulated/chains/parachains/coretime/coretime-westend" }
coretime-westend-runtime = { path = "cumulus/parachains/runtimes/coretime/coretime-westend" }
```

File: substrate/frame/bags-list/src/list/tests.rs [hunk 4]
```diff
@@ -777,7 +777,8 @@ mod bags {
 			assert_eq!(bag_1000.tail, Some(4));
 			assert_eq!(bag_1000.iter().count(), 3);
 			bag_1000.insert_node_unchecked(node(4, None, None, bag_1000.bag_upper)); // panics in debug
-			assert_eq!(bag_1000.iter().count(), 3); // in release we expect it to silently ignore the request.
+			assert_eq!(bag_1000.iter().count(), 3); // in release we expect it to silently ignore the
+			                            // request.
```
Comment (acatangiu) [inline substrate/frame/bags-list/src/list/tests.rs:781]: change unrelated to this PR, would not include

File: substrate/frame/nis/src/lib.rs [hunk 5]
```diff
@@ -755,7 +755,13 @@ pub mod pallet {
...
 					// dust and the beneficiary account doesn't exist.
 					.or_else(
-						|e| if e == TokenError::CannotCreate.into() { Ok(()) } else { Err(e) },
+						|e| {
+							if e == TokenError::CannotCreate.into() {
+								Ok(())
+							} else {
+								Err(e)
+							}
+						},
```
Comment (acatangiu) [inline substrate/frame/nis/src/lib.rs:764]: change unrelated to this PR, would not include

**Review by acatangiu — COMMENTED** [review]
very nice, thanks @programskillforverification !

everything looks good, feel free to do the same for Westend.

Could you please also include the teleport tests for all chains that don't have it? (e.g. coretime emulated chains you just added)

File: substrate/frame/bags-list/src/list/tests.rs (hunk 4 above)
Comment (programskillforverification) [reply substrate/frame/bags-list/src/list/tests.rs:781]: Maybe due to merge updated master, and push forcefully.

File: Cargo.toml (hunk 3 above)
Comment (programskillforverification) [reply Cargo.toml:676]: Yeah. I focus on rococo and not care westend

**Comment by programskillforverification** [comment]
> very nice, thanks @programskillforverification !
> 
> everything looks good, feel free to do the same for Westend.
> 
> Could you please also include the teleport tests for all chains that don't have it? (e.g. coretime emulated chains you just added)

What should the teleport tests include in coretime? Teleport tests looks different for different chains.

**Comment by acatangiu** [comment]
> What should the teleport tests include in coretime? Teleport tests looks different for different chains.

Basic ROC (on Rococo) and WND (on Westend) teleport tests: between Coretime and Relay and between coretime and other system chains.

(this ^ should be minimum for all system chains)

**Comment by programskillforverification** [comment]
> > What should the teleport tests include in coretime? Teleport tests looks different for different chains.
> 
> Basic ROC (on Rococo) and WND (on Westend) teleport tests: between Coretime and Relay and between coretime and other system chains.
> 
> (this ^ should be minimum for all system chains)

Now, macro `test_relay_is_trusted_teleporter!` is already added.

```rust
/// Limited Teleport of native asset from Relay Chain to the System Parachain should work
#[test]
fn limited_teleport_native_assets_from_relay_to_system_para_works() {
	// Init values for Relay Chain
	let amount_to_send: Balance = WESTEND_ED * 1000;
	let dest = Westend::child_location_of(AssetHubWestend::para_id());
	let beneficiary_id = AssetHubWestendReceiver::get();
	let test_args = TestContext {
		sender: WestendSender::get(),
		receiver: AssetHubWestendReceiver::get(),
		args: TestArgs::new_relay(dest, beneficiary_id, amount_to_send),
	};

	let mut test = RelayToSystemParaTest::new(test_args);
...
```

Is this test function outdated? If so, I want to rewrite all teleport tests using macro.

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
@franciscoaguirre PTAL

@programskillforverification are you part of [Polkadot Fellowship](https://collectives.subsquare.io/fellowship/members)? If not, please add a Polkadot address so we can provide a tip for this useful work 👍

File: cumulus/parachains/integration-tests/emulated/common/src/macros.rs [hunk 6]
```diff
@@ -130,3 +130,279 @@ macro_rules! test_parachain_is_trusted_teleporter {
 		}
 	};
 }
+
+#[macro_export]
```
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/common/src/macros.rs:134]: We need to publish a new version of `emulated-integration-tests-common` to be able to use this in the fellowship/runtimes repo.

@programskillforverification please also add a [prdoc](https://github.com/paritytech/polkadot-sdk/tree/master/prdoc) with short changelog description and include at least `emulated-integration-tests-common` (bump minor) there.

Comment (programskillforverification) [reply cumulus/parachains/integration-tests/emulated/common/src/macros.rs:134]: > We need to publish a new version of `emulated-integration-tests-common` to be able to use this in the fellowship/runtimes repo.
> 
> @programskillforverification please also add a [prdoc](https://github.com/paritytech/polkadot-sdk/tree/master/prdoc) with short changelog description and include at least `emulated-integration-tests-common` (bump minor) there.

I will add it asap

**Comment by programskillforverification** [comment]
> @franciscoaguirre PTAL
> 
> @programskillforverification are you part of [Polkadot Fellowship](https://collectives.subsquare.io/fellowship/members)? If not, please add a Polkadot address so we can provide a tip for this useful work 👍

Thanks a lot. I will add my address.

**Comment by bkontur** [comment]
I think we should also reflect here upcoming changes for dry-run api from this PR: https://github.com/polkadot-fellows/runtimes/pull/380/files#diff-7346706278168c97dfec9e73bec1a115274a2be89a3147b914dd2eedec9f6269R57

**Comment by programskillforverification** [comment]
> I think we should also reflect here upcoming changes for dry-run api from this PR: https://github.com/polkadot-fellows/runtimes/pull/380/files#diff-7346706278168c97dfec9e73bec1a115274a2be89a3147b914dd2eedec9f6269R57

I am refactoring teleport in a new PR, which covers some part of this PR. For XcmPaymentApi and DryRunApi, maybe are in separate PR.

**Review by acatangiu — APPROVED** [review]
Changes are good, you just have to make the CI happy - you can check the failed checks on their `details` button to find out what/why's failing

**Comment by programskillforverification** [comment]
> Changes are good, you just have to make the CI happy - you can check the failed checks on their `details` button to find out what/why's failing

I fixed most except `check-semver`. The rust version need to be upgraded in `check-semver` to run it.

**Comment by acatangiu** [comment]
@bkontur @franciscoaguirre please review

File: prdoc/pr_4930.prdoc [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Add test macro to emulated chains
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      This PR add an portable test macro for all emulated-chains runtimes and chains could
+      claim error for unknown asset.
```
Comment (franciscoaguirre) [inline prdoc/pr_4930.prdoc:10]:
```suggestion
      This PR adds a portable test macro that can be used to test trapped assets can be
      claimed in an emulated chain.
```

File: prdoc/pr_4930.prdoc [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+
+crates:
```
Comment (franciscoaguirre) [inline prdoc/pr_4930.prdoc:13]: Check if you need to bump more crates. Although most crates have publish = false, so they don't need to be included

**Review by franciscoaguirre — APPROVED** [review]
Thank you for this effort!

File: cumulus/parachains/integration-tests/emulated/chains/parachains/coretime/coretime-rococo/src/genesis.rs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+use sp_core::storage::Storage;
+
+// Cumulus
+use emulated_integration_tests_common::{
+	accounts, build_genesis_storage, collators, SAFE_XCM_VERSION,
+};
+use parachains_common::Balance;
+
+pub const PARA_ID: u32 = 1005;
+pub const ED: Balance = testnet_parachains_constants::rococo::currency::EXISTENTIAL_DEPOSIT;
```
[the remark this answers is not in the document]
Comment (programskillforverification) [reply cumulus/parachains/integration-tests/emulated/chains/parachains/coretime/coretime-rococo/src/genesis.rs:26]: > Is this the ED for the relay? If so, it should be smaller for system parachains

What is "for the relay"? I just know it is set to 1/10 of its parent relay chain.

**Review by acatangiu — APPROVED** [review]

**Pull request merged**